noir-projects/aztec-nr/aztec/src/macros/notes/mod.nr

```diff
@@ -560,43 +637,123 @@ comptime fn generate_finalization_payload(
+            setup_log_slot: Field,
+            public_values: [Field; $public_values_length],
         }
 
         impl $finalization_payload_name {
-            fn new(mut self, hiding_point: aztec::protocol_types::point::Point, $args) -> $finalization_payload_name {
+            fn new(mut self, context: &mut aztec::prelude::PublicContext, slot: Field, $args) -> $finalization_payload_name {
+                self.context = context;
+
+                self.hiding_point_slot = slot;
+                self.setup_log_slot = slot + aztec::protocol_types::point::POINT_LENGTH as Field;
+
                 $aux_vars_for_serialization
-                self.log = [$fields];
+                self.public_values = [$fields];
+
+                self
+            }
+
+            fn emit(self) {
+                self.emit_note_hash();
+                self.emit_log();
+            }
+
+            fn emit_note_hash(self) {
+                // Read the hiding point from "transient" storage and check it's not empty to ensure the transfer was prepared
+                let hiding_point: aztec::prelude::Point = self.context.storage_read(self.hiding_point_slot);
+                assert(!aztec::protocol_types::traits::is_empty(hiding_point), "transfer not prepared");
 
                 $msm_aux_vars
                 let finalization_hiding_point = std::embedded_curve_ops::multi_scalar_mul(
                     [$generators],
                     [$scalars]
                 ) + hiding_point;
 
-                self.note_hash = finalization_hiding_point.x;
-                self
+                let note_hash = finalization_hiding_point.x;
+
+                self.context.push_note_hash(note_hash);
+
+                // We reset public storage to zero to achieve the effect of transient storage - kernels will squash
+                // the writes
+                // TODO(#9376): Uncomment the following line.
+                // self.context.storage_write(self.hiding_point_slot, [0; aztec::protocol_types::point::POINT_LENGTH]);
+            }
+
+            fn emit_log(self) {
+                // We load the setup log from storage
+                let setup_log_fields: [Field; $setup_log_field_length] = self.context.storage_read(self.setup_log_slot);
+
+                // We convert the lof from fields to bytes
+                let setup_log: [u8; $setup_log_byte_length] = aztec::utils::bytes::fields_to_bytes(setup_log_fields);
+
+                // First byte should be zero as the number of public values is populated in public
+                assert(setup_log[0] == 0, "Non-zero first byte");
+
+                // We append the public value to the log and emit it as unencrypted log
+                let mut finalization_log = [0; $finalization_log_byte_length];
+
+                // Iterate over the partial log and copy it to the final log
+                for i in 1..setup_log.len() {
+                    finalization_log[i] = setup_log[i];
+                }
```

Comment:
Why do we need the first byte in `setup_log` to be zero if we're going to be skipping it here?

Reply:
It's only a sanity check that we did not mess something up. Realized that I unnecessarily cluttered the code with the first byte. Will clean that up.